server/src/main/java/timely/store/iterators/LookaheadIterator.java

```diff
@@ -0,0 +1,148 @@
+package timely.store.iterators;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+
+/**
+ * Accumulo iterator that wraps another iterator and implements a lookahead
+ * feature. As you are iterating over the keys and values, you can peek ahead to
+ * the next key/value. Be aware that calling {@link #next()} after
+ * {@link #peek()} will return the same key and value. Additionally, you must
+ * consume both the key and value after calling {@link #next()} to clear the
+ * internal state. Example usage:
+ * 
+ * <pre>
+ * LookaheadIterator iter = new LookaheadIterator(source);
+ * iter.seek();
+ * iter.hasTop();
+ * iter.getTopKey();
+ * iter.getTopValue();
+ * 
+ * // Lookahead at the next key/value
+ * KeyValuePair lookahead = iter.peek();
+ * lookahead.getKey();
+ * lookahead.getValue();
+ * 
+ * // Move ahead to the next value, which is really the last value because of the
+ * // peek.
+ * iter.next();
+ * iter.hasTop();
+ * iter.getTopKey();
+ * iter.getTopValue(); // must call getTopKey and getTopValue after a peek to clear
+ * // the internal state
+ * 
+ * iter.next();
+ * iter.hasTop();
+ * iter.getTopKey();
+ * iter.getTopValue();
+ * </pre>
+ */
```

Comment:
You don't know, at the time of you next if you will trigger a new block to be read. If the currBlock in The LocalityGroupReader is nearing its end you may be wasting a lot of time doing this next only to be torn down and recreated from that block.